### PR TITLE
Deserialization is now thread-safe 

### PR DESCRIPTION
**Pull request by torbacz**
## Description

- Removed all static types with class scope from deserialize method.

## Motivation and Context

- Resolves nanoFramework/Home#1160

## How Has This Been Tested?

- Provided coded deployed on ESP32 and run for about 1 minute (before changes it throws error after ~20s)

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [x] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [ ] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [ ] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/master/CONTRIBUTING.md) document.
- [ ] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

### CODE REVIEW

**Comment by torbacz** [comment]
I've tried to create unit test for that, but failed 😞 Looks like on emulator it's working fine all the time.

**Comment by torbacz** [comment]
Performance results
image

All results within 10ms should be ignored because stopwatch in ESP32 has 10 ms accuracy. For example, I've changed only Deserialize method, yet string serialization is 8% faster? When I run benchmark again it's only 2%. I would call it "benchmark noise" 😄

File: nanoFramework.Json/JsonConvert.cs [hunk 1]
```diff
@@ -90,8 +90,8 @@ public static object DeserializeObject(StreamReader dr, Type type)
         private static object Deserialize(StreamReader dr)
         {
             // Read the DataReader into jsonBytes[]
-            jsonBytes = new byte[dr.BaseStream.Length];
-            jsonPos = 0;
+            var jsonBytes = new byte[dr.BaseStream.Length];
```
Comment (Ellerbach) [inline nanoFramework.Json/JsonConvert.cs:93]: hum, where were they defined before the change?

**Comment by Ellerbach** [comment]
> I've tried to create unit test for that, but failed 😞 Looks like on emulator it's working fine all the time.

We're looking at how to run properly tests on real hardware. We've done some exploration, that's possible, just not that super easy and requires some infrastructure. But what you can do is write the Unit Tests anyway, run them on a real device as well and if they always work in the Win32 nanoCLR, then that's fine as well.

File: nanoFramework.Json/JsonConvert.cs (hunk 1 above)
Comment (torbacz) [reply nanoFramework.Json/JsonConvert.cs:93]: In class as static field. Check line 768-769 in old file.

Comment (Ellerbach) [reply nanoFramework.Json/JsonConvert.cs:93]: so then they should be removed if you instantiate them locally every time?

Comment (torbacz) [reply nanoFramework.Json/JsonConvert.cs:93]: They are, that's why I wrote "768-769 in **old file**" ;)

**Comment by torbacz** [comment]
> > I've tried to create unit test for that, but failed 😞 Looks like on emulator it's working fine all the time.
> 
> We're looking at how to run properly tests on real hardware. We've done some exploration, that's possible, just not that super easy and requires some infrastructure. But what you can do is write the Unit Tests anyway, run them on a real device as well and if they always work in the Win32 nanoCLR, then that's fine as well.

Test included.

**Comment by torbacz** [comment]
Ready for merge.

File: nanoFramework.Json.Test/JsonThreadSafeTests.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                var newThread = new Thread(() => ThreadMethod(threadLoggger));
+                newThread.Start();
+                threads.Add(newThread);
+                Thread.Sleep(TimeSpan.FromSeconds(1));
+            }
+
+            // Let's wait a little bit so threads can do their job
+            Thread.Sleep(TimeSpan.FromSeconds(30));
+
+            // Abrot all threads
```
Comment (josesimoes) [inline nanoFramework.Json.Test/JsonThreadSafeTests.cs:44]:
```suggestion
            // Abort all threads
```

**Comment by josesimoes** [comment]
@torbacz can you please address Sonarcloud code smell please?

**Comment by josesimoes** [comment]
> @josesimoes done my best, 2 issues left but nothing I can do with them, json lib inside is using set.

No worries about these!

**Pull request merged**